This PR adds the abstract class DQ_CoppeliaSimInterface

juanjqo opened the pull request:
![](https://img.shields.io/badge/type-high_priority-red)![](https://img.shields.io/badge/status-experimental-red)![](https://img.shields.io/badge/warning-internal_modifications-yellow)

Hi @dqrobotics/developers,

This PR adds the abstract class `DQ_CoppeliaSimInterface`. This class was originally proposed and discussed [here](https://github.com/dqrobotics/matlab/pull/112), and it is based on the [current C++ implementation](https://github.com/dqrobotics/cpp-interface-coppeliasim). This abstract class will enable the implementation of the concrete class `DQ_CoppeliaSimInterfaceZMQ`, which is under development.

# UML:
<img src="https://github.com/user-attachments/assets/075448f9-35ac-43fe-8666-4fe5f74f0ef4" alt="drawing" width="150"/>

# Updated UML:
<img src="https://github.com/user-attachments/assets/ed13a3a0-f752-4ec9-9bcd-db2e9abff820" alt="drawing" width="400"/>


Kind regards, 

Juancho

juanjqo committed:
Added the DQ_CoppeliaSimInterface abstract class

juanjqo committed:
[DQ_CoppeliaSimInterface] updated the signature of the connect method.

juanjqo committed:
[DQ_CoppeliaSimInterface.m] Updated the name of the argument in the methods get_object_handle and get_object_handles to match the C++ version.

juanjqo committed:
Update README.md

bvadorno commented:
Thanks, @juanjqo. The class definition looks good to me.

@ffasilva, could you please review the code?

Many thanks,
Bruno

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+
+% (C) Copyright 2025 DQ Robotics Developers
```
As per @bvadorno's [comment](https://github.com/dqrobotics/matlab/pull/108/files#r1718378842) in previous pull requests, please amend the copyright date.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+
+classdef  (Abstract) DQ_CoppeliaSimInterface < handle
+    
+    
```
Minor comment, but other MATLAB classes don't seem to have this indenting pattern of double space between `classdef` and `methods`. Maybe you could consider removing them.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+    methods(Abstract)
+            % This method connects to CoppeliaSim.
+            % Calling this function is required before anything else can happen.
+            status = connect(obj, host, port, TIMEOUT_IN_MILISECONDS);
```
Why is the last argument in all caps? Is it meant to be a constant somehow?

juanjqo replied:
@ffasilva I copied some styles from low-level implementation of the legacy API in C++. But I don't have a specific reason to use capital letters in the last argument. I changed it to lowercase.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            status = connect(obj, host, port, TIMEOUT_IN_MILISECONDS);
+    
+            % This method enables or disables the stepped (synchronous) mode
+            % for the remote API server service that the client is connected to.
```
I think it would be better to rephrase to: "This method enables or disables the stepped (synchronous) mode for the remote API server service to which the client is connected."

ffasilva replied:
@juanjqo, did you forget to change this or do you prefer the original phrasing?

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+    
+            % This method gets the handles for a cell array of 
+            % object names in the the CoppeliaSim scene.
+            handles = get_object_handles(obj,objectnames);
```
Minor comment as well, but for some methods you used a space between arguments, e.g.,

`status = connect(obj, host, port, TIMEOUT_IN_MILISECONDS);`

and for others you did not, e.g.,

`handles = get_object_handles(obj,objectnames);`

Maybe you could establish a convention.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            % This method sets the joint positions of a robot in the CoppeliaSim scene.        
+            set_joint_positions(obj,jointnames,joint_positions);
+    
+            % This method sets the joint positions of a robot in the CoppeliaSim scene. 
```
gets*

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            set_object_pose(obj,objectname,pose);
+    
+            % This method sets the joint positions of a robot in the CoppeliaSim scene.        
+            set_joint_positions(obj,jointnames,joint_positions);
```
I'm not sure what would be the best option, but we started to adopt the convention of `get/set_configuration` (extended to `get/set_configuration_velocities`, `get/set_configuration_space_torques`, `set_target_configuration`, etc.) when we implemented [DQ_SerialVrepRobot.m](https://github.com/dqrobotics/matlab/blob/master/interfaces/vrep/DQ_SerialVrepRobot.m). Should we follow the same nomenclature here? (for reference, this was discussed on several points in this pull [request](https://github.com/dqrobotics/matlab/pull/108#discussion_r1718392079)).

juanjqo replied:
@ffasilva I think for this class the use of  `{set,get}_joint_xxx` is suitable, since it is a high-level class. Consider the youBot holonomic platform, which in CoppeliaSim is equipped with 9 joints (5 free joints + 4 velocity-commanded joints). If the user wants to read or command those joints using methods as `{set,get}_configuration` would be weird (at least in the class DQ_CoppeliaSimInterface), since the robot configuration in this case is given by q=[x,y, phi].

ffasilva replied:
Hi, @juanjqo. I see what you mean. These methods are meant for low-level joint information, and the robot classes can have the high-level methods for configurations. It does make sense to keep the names then!

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+    
+            % This method sets the joint target positions of a robot in the CoppeliaSim scene. 
+            % It is required a dynamics enabled scene, and joints in dynamic mode 
+            % with position control mode. 
```
I believe it would be better as "It requires a dynamics-enabled scene and joints in dynamic mode with position control mode."

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            % This method sets the joint target positions of a robot in the CoppeliaSim scene. 
+            % It is required a dynamics enabled scene, and joints in dynamic mode 
+            % with position control mode. 
+            % information about joint modes:
```
"For more information about joint modes:"

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            % This method sets the joint target velocities of a robot in the CoppeliaSim scene.
+            % It is required a dynamics enabled scene, and joints in dynamic mode 
+            % with velocity control mode. Check this link for more
+            % information about joint modes:
```
Same comments as in `set_joint_target_positions`.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,111 @@
+            % This method sets the joint torques of a robot in the CoppeliaSim scene.
+            % It is required a dynamics enabled scene, and joints in dynamic mode 
+            % with velocity or force/torque control mode. Check this link for more
+            % information about joint modes:
```
Same comments as in `set_joint_target_positions`.

ffasilva reviewed (commented):
Hi @juanjqo,

Thank you for the pull request! Most of my comments are cosmetic, but the nomenclatures of the methods for getting and setting joint positions, velocities, and torques might require some discussion.

Kind regards,
Frederico

juanjqo committed:
[DQ_CoppeliaSimInterface] Implemented all the style modifications proposed by Frederico. This includes the use of lowercase style for all arguments in the methods, and keep the same spacing between them.

juanjqo committed:
Merge branch 'dev' of https://github.com/juanjqo/matlab-interface-coppeliasim into dev

juanjqo committed:
[DQ_CoppeliaSimInterface] Updated the documentation of the method get_joint_positions to match the style of the other methods.

juanjqo committed:
[DQ_CoppeliaSimInterface] Renamed the methods set,get_joint_torques to set_joint_target_forces and get_joint_forces to comply the style of the set/get methods related to joint positions, and velocities.

ffasilva commented on README.md:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+# matlab-interface-coppeliasim
+
+Contains the common interfaces used to interface DQ Robotics with CoppeliaSim
```
The "interfaces used to interface" sounds a bit strange. Maybe we could use: "Abstract class defining the DQ Robotics interface with CoppeliaSim in MATLAB." What do you think?

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,106 @@
+
```
A wild lost extra space appears.

ffasilva commented on interfaces/coppeliasim/DQ_CoppeliaSimInterface.m:
```diff
@@ -0,0 +1,106 @@
+            set_joint_target_positions(obj, jointnames, joint_target_positions);
+    
+            % This method gets the joint velocities in the CoppeliaSim scene. 
+            joint_velocities = get_joint_velocities(obj,jointnames);
```
Missing space between arguments:

```matlab
joint_velocities = get_joint_velocities(obj, jointnames);
```

ffasilva reviewed (commented):
Hi @juanjqo,

I've added some more minor comments, and I agree with your reasoning for keeping the `set/get_joint` information name convention for the methods.

I believe we're pretty much good to go here.

Kind regards,
Frederico

juanjqo committed:
[DQ_CoppeliaSimInterface, README] Remove extra spaces and updated the readme file.

juanjqo commented:
> Hi @juanjqo,
> 
> I've added some more minor comments, and I agree with your reasoning for keeping the `set/get_joint` information name convention for the methods.
> 
> I believe we're pretty much good to go here.
> 
> Kind regards, Frederico

Thanks, @ffasilva. I updated the PR to take your comments into account.

ffasilva commented:
Hi @juanjqo,

Thank you! I have nothing else to add.

Kind regards,
Frederico

ffasilva commented:
Hi @bvadorno,

We finished the review. Shall we approve this PR?

Kind regards,
Frederico
